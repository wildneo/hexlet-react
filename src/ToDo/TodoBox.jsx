import axios from 'axios';
import React from 'react';
import update from 'immutability-helper';
import routes from './routes';
import Item from './Item';

export default class TodoBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      tasks: [],
      text: '',
    };

    this.handleChange = ({ target: { value } }) => {
      this.setState({ text: value });
    };

    this.handleSubmit = (event) => {
      event.preventDefault();
      const { text } = this.state;
      this.addTask(text);
    };

    this.handleClick = (id) => (event) => {
      event.preventDefault();
      this.toggleTaskState(id);
    };
  }

  componentDidMount() {
    this.getTasks();
  }

  async getTasks() {
    const { data } = await axios.get(routes.tasksPath());
    this.setState({ tasks: data });
  }

  async addTask(text) {
    const { tasks } = this.state;
    const { data } = await axios.post(routes.tasksPath(), { text });
    this.setState({ text: '', tasks: update(tasks, { $push: [data] }) });
  }

  async toggleTaskState(id) {
    const { tasks } = this.state;
    const index = tasks.findIndex((t) => t.id === id);
    const task = tasks[index];
    const url = (t) => {
      switch (t.state) {
        case 'active':
          return routes.finishTaskPath(t.id);

        case 'finished':
          return routes.activateTaskPath(t.id);

        default:
          throw new Error(`Unexpected state ${t.state}`);
      }
    };
    const { data } = await axios.patch(url(task));
    this.setState({ tasks: update(tasks, { [index]: { $set: data } }) });
  }

  renderForm() {
    const { text } = this.state;

    return (
      <form className="todo-form form-inline mx-3" onSubmit={this.handleSubmit}>
        <div className="form-group">
          <input
            className="form-control mr-3"
            placeholder="I am going..."
            type="text"
            value={text}
            onChange={this.handleChange}
            required
          />
        </div>
        <button type="submit" className="btn btn-primary">add</button>
      </form>
    );
  }

  renderTasks(taskState) {
    const { tasks } = this.state;
    const filtered = tasks.filter(({ state }) => state === taskState);

    if (filtered.length === 0) {
      return null;
    }

    return (
      <div className={`todo-${taskState}-tasks`}>
        {filtered.map((task) => <Item key={task.id} task={task} onClick={this.handleClick} />)}
      </div>
    );
  }

  render() {
    return (
      <div>
        <div className="mb-3">
          {this.renderForm()}
        </div>
        {this.renderTasks('active')}
        {this.renderTasks('finished')}
      </div>
    );
  }
}
