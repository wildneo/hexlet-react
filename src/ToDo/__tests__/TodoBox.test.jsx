import nock from 'nock';
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import timeout from 'timeout-then';
import TodoBox from '../TodoBox';

Enzyme.configure({ adapter: new Adapter() });
axios.defaults.adapter = httpAdapter;
nock.disableNetConnect();

const host = 'http://localhost';

test('TodoBox 1', async () => {
  nock(host)
    .get('/tasks')
    .reply(200, []);

  const wrapper = mount(<TodoBox />);
  const input = wrapper.find('input');
  const form = wrapper.find('form');

  await timeout(100);

  expect(wrapper.render()).toMatchSnapshot();
  input.simulate('change', { target: { value: 'new task' } });

  nock(host)
    .post('/tasks', {
      text: 'new task',
    })
    .reply(200, { id: 1, state: 'active', text: 'new task' });
  form.simulate('submit');

  await timeout(100);

  expect(wrapper.render()).toMatchSnapshot();
});

test('TodoBox 2', async () => {
  const tasks = [
    { id: 2, text: 'task 2', state: 'finished' },
    { id: 1, text: 'task 1', state: 'active' },
  ];
  nock(host)
    .get('/tasks')
    .reply(200, tasks);

  const wrapper = mount(<TodoBox />);

  await timeout(100);

  expect(wrapper.render()).toMatchSnapshot();

  wrapper.update();

  const activeTask = wrapper.find('.todo-active-tasks .todo-task');
  nock(host)
    .patch(`/tasks/${tasks[1].id}/finish`)
    .reply(200, { ...tasks[1], state: 'finished' });
  activeTask.simulate('click');

  await timeout(100);

  wrapper.update();

  expect(wrapper.render()).toMatchSnapshot();

  const finishedTask = wrapper.find('.todo-task').at(0);
  nock(host)
    .patch(`/tasks/${tasks[0].id}/activate`)
    .reply(200, { ...tasks[0], state: 'active' });
  finishedTask.simulate('click');

  await timeout(100);

  wrapper.update();

  expect(wrapper.render()).toMatchSnapshot();
});
