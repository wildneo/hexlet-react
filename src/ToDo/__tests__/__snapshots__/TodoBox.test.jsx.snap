// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoBox 1 1`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form form-inline mx-3"
    >
      <div
        class="form-group"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
</div>
`;

exports[`TodoBox 1 2`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form form-inline mx-3"
    >
      <div
        class="form-group"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          new task
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 2 1`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form form-inline mx-3"
    >
      <div
        class="form-group"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          task 1
        </a>
      </div>
    </div>
  </div>
  <div
    class="todo-finished-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        2
      </div>
      <div
        class="col"
      >
        <s>
          <a
            class="todo-task"
            href="#"
          >
            task 2
          </a>
        </s>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 2 2`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form form-inline mx-3"
    >
      <div
        class="form-group"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-finished-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        2
      </div>
      <div
        class="col"
      >
        <s>
          <a
            class="todo-task"
            href="#"
          >
            task 2
          </a>
        </s>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <s>
          <a
            class="todo-task"
            href="#"
          >
            task 1
          </a>
        </s>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 2 3`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form form-inline mx-3"
    >
      <div
        class="form-group"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        2
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          task 2
        </a>
      </div>
    </div>
  </div>
  <div
    class="todo-finished-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <s>
          <a
            class="todo-task"
            href="#"
          >
            task 1
          </a>
        </s>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 2 4`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form form-inline mx-3"
    >
      <div
        class="form-group"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-finished-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        2
      </div>
      <div
        class="col"
      >
        <s>
          <a
            class="todo-task"
            href="#"
          >
            task 2
          </a>
        </s>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <s>
          <a
            class="todo-task"
            href="#"
          >
            task 1
          </a>
        </s>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoBox 2 5`] = `
<div>
  <div
    class="mb-3"
  >
    <form
      class="todo-form form-inline mx-3"
    >
      <div
        class="form-group"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        add
      </button>
    </form>
  </div>
  <div
    class="todo-active-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        2
      </div>
      <div
        class="col"
      >
        <a
          class="todo-task"
          href="#"
        >
          task 2
        </a>
      </div>
    </div>
  </div>
  <div
    class="todo-finished-tasks"
  >
    <div
      class="row"
    >
      <div
        class="col-1"
      >
        1
      </div>
      <div
        class="col"
      >
        <s>
          <a
            class="todo-task"
            href="#"
          >
            task 1
          </a>
        </s>
      </div>
    </div>
  </div>
</div>
`;
